go-controller/pkg/util/kube.go

```diff
@@ -220,6 +250,72 @@ func newKubernetesRestConfig(conf *config.KubernetesConfig) (*rest.Config, error
 	return kconfig, nil
 }
 
+// startNodeCertificateManager manages the creation and rotation of the node-specific client certificate.
+// When there is no existing certificate, it will use the BootstrapKubeconfig kubeconfig to create a CSR and it will
+// wait for the certificate before returning.
+func startNodeCertificateManager(conf *config.KubernetesConfig, defaultKConfig *rest.Config) error {
+	nodeName := os.Getenv("K8S_NODE")
+	if nodeName == "" {
+		return fmt.Errorf("failed to get the node name required for the certificate from K8S_NODE env")
+	}
+
+	bootstrapKConfig, err := clientcmd.BuildConfigFromFlags("", conf.BootstrapKubeconfig)
+	if err != nil {
+		return fmt.Errorf("failed to load bootstrap kubeconfig from %s, err: %v", conf.BootstrapKubeconfig, err)
+	}
+	// If we have a valid certificate, use that to fetch CSRs.
+	// Otherwise, use the bootstrap credentials.
+	// https://github.com/kubernetes/kubernetes/blob/068ee321bc7bfe1c2cefb87fb4d9e5deea84fbc8/cmd/kubelet/app/server.go#L953-L963
+	newClientsetFn := func(current *tls.Certificate) (kubernetes.Interface, error) {
+		cfg := bootstrapKConfig
+		if current != nil {
+			cfg = defaultKConfig
+		}
+		return kubernetes.NewForConfig(cfg)
+	}
+
+	certificateStore, err := certificate.NewFileStore(certNamePrefix, conf.CertDir, conf.CertDir, "", "")
+	if err != nil {
+		return fmt.Errorf("failed to initialize the certificate store: %v", err)
+	}
+	certManager, err := certificate.NewManager(&certificate.Config{
+		ClientsetFn: newClientsetFn,
+		Template: &x509.CertificateRequest{
+			Subject: pkix.Name{
+				CommonName:   fmt.Sprintf("%s:%s", certCommonNamePrefix, nodeName),
+				Organization: []string{certOrganization},
+			},
+		},
+		RequestedCertificateLifetime: &conf.CertDuration,
+		SignerName:                   certificatesv1.KubeAPIServerClientSignerName,
+		Usages:                       certUsages,
+		CertificateStore:             certificateStore,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to initialize the certificate manager: %v", err)
+	}
+
+	if conf.CertDuration < time.Hour {
+		// the default value for CertCallbackRefreshDuration (5min) is too long for short-lived certs,
+		// set it to a more sensible value
+		transport.CertCallbackRefreshDuration = time.Second * 10
+	}
+	certManager.Start()
```

Comment:
I am a bit uneasy about two things:

- No controlled shutdown of the certManager
- Multiple certManager can be started if NewKubernetesClientset, which seems like a mere utility function (vene if right now its called only once), is called multiple times.

So should this certManager be started from a more suitable place?

Reply:
I think I initially included it here due to some dependency but this is no longer required.
I moved it out as a exported method and call it from `startOvnKube` with a context.